lib/matplotlib/tests/test_animation.py

```diff
@@ -261,3 +262,86 @@ def test_failing_ffmpeg(tmpdir, monkeypatch):
                 make_animation().save("test.mpeg")
     finally:
         animation.writers.reset_available_writers()
+
+
+def test_funcanimation_holds_frames_by_default():
+    fig, ax = plt.subplots()
+    line, = ax.plot([], [])
+
+    class Frame(dict):
+        # this subclassing enables to use weakref.ref()
+        pass
+
+    def init():
+        line.set_data([], [])
+        return line,
+
+    def animate(frame):
+        line.set_data(frame['x'], frame['y'])
+        return line,
+
+    frames_generated = []
+
+    def frames_generator():
+        for _ in range(5):
+            x = np.linspace(0, 10, 100)
+            y = np.random.rand(100)
+
+            frame = Frame(x=x, y=y)
+
+            # collect weak references to frames
+            # to validate their references later
+            frames_generated.append(weakref.ref(frame))
+
+            yield frame
+
+    anim = animation.FuncAnimation(fig, animate, init_func=init,
+                                   frames=frames_generator, save_count=5)
+
+    writer = NullMovieWriter()
+    anim.save('unused.null', writer=writer)
+
+    for f in frames_generated:
+        assert f() is not None
+
+
+def test_funcanimation_does_not_hold_frames_when_cache_frame_data_is_False():
```

Comment:
could use pytest parametrization to avoid duplicating nearly the entire code.

Reply:
@anntzer Thanks for the hint.
I've applied your suggestion in the rebased commit.